codaco/Network-Canvas
# Update redux for protocols & sessions

Resolves #545. Resolves #558.

This updates redux actions (with some minor changes to state shape) to better support protocol exporting and simplify some of the import/load/set flows with sessions. This builds on the FinishSession interface branch (a merge should wait on #559), but the two can be reviewed in parallel.

This also adds support for loading devtools extensions in electron (as redux-devtools was very useful for this effort), but installation left as a manual exercise (see Readme).

### Notes on changes
- Removes some of the differences between `loadFactoryProtocol` and `loadProtocol`
  + In particular, both now make use of `addSession()`
  + ...and session IDs are no longer generated by protocol actions
  + This reduces some of the coupling, but interfaces must call addSession in both cases now. (Unwanted side-effects from programming error should be solved by #570.)
- Removes the `remoteId` that was passed through actions (to provide initial support for export). Remote IDs are now calculated locally as they are not server-specific.

### Manual testing

We don't have much integration testing yet, and there's still integration between sessions and protocols. I've gone through combinations of {protocolType} + {loadAction|exportAction}:

- Factory protocols vs Downloaded (Server- and URL-sourced) protocols
- Loading
  - existing sessions
    - via app reload
    - via 'resume interview'
  - new Sessions
    - via protocol cards (startup screen)
    - via download flows
- Exporting (or not) as expected
